File: rest-api/src/main/java/com/kt/kol/api/user/repository/UserRepository.java
package com.kt.kol.api.user.repository;

import com.kt.kol.api.user.model.User;
import org.springframework.data.r2dbc.repository.Query;
import org.springframework.data.repository.reactive.ReactiveCrudRepository;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Repository
public interface UserRepository extends ReactiveCrudRepository<User, Long> {

    Mono<User> findByUsername(String username);

    Mono<User> findByEmail(String email);

    Flux<User> findByActive(Boolean active);

    @Query("SELECT * FROM users WHERE first_name LIKE :name% OR last_name LIKE :name%")
    Flux<User> searchByName(String name);

    Mono<Boolean> existsByUsername(String username);

    Mono<Boolean> existsByEmail(String email);
}

File: rest-api/src/main/kotlin/com/kt/kol/api/user/repository/UserRepository.kt
package com.kt.kol.api.user.repository

import com.kt.kol.api.user.model.User
import kotlinx.coroutines.flow.Flow
import org.springframework.data.r2dbc.repository.Query
import org.springframework.data.repository.kotlin.CoroutineCrudRepository
import org.springframework.stereotype.Repository

@Repository
interface UserRepository : CoroutineCrudRepository<User, Long> {

    suspend fun findByUsername(username: String): User?

    @Query("SELECT * FROM users ORDER BY id ASC LIMIT :size OFFSET :offset")
    fun findPage(offset: Long, size: Int): Flow<User>

    @Query("SELECT COUNT(*) FROM users")
    suspend fun countAll(): Long

    @Query("SELECT * FROM users WHERE first_name LIKE :name% OR last_name LIKE :name% ORDER BY id ASC LIMIT :size OFFSET :offset")
    fun searchByNamePage(name: String, offset: Long, size: Int): Flow<User>

    @Query("SELECT COUNT(*) FROM users WHERE first_name LIKE :name% OR last_name LIKE :name%")
    suspend fun countByNameSearch(name: String): Long

    suspend fun existsByUsername(username: String): Boolean

    suspend fun existsByEmail(email: String): Boolean
}